File: packages/node/src/crypto.ts
import {
  createHmac,
  hkdfSync,
  randomBytes,
  createCipheriv,
  createDecipheriv,
  timingSafeEqual,
} from 'node:crypto';

/** base64url-encode without padding (JWT segment encoding). */
export function base64url(input: Buffer | string): string {
  return Buffer.from(input).toString('base64url');
}

export function hmacSha256(data: string, secret: string): Buffer {
  return createHmac('sha256', secret).update(data).digest();
}

/** Constant-time comparison of two base64url signatures. */
export function safeEqual(a: string, b: string): boolean {
  const ba = Buffer.from(a);
  const bb = Buffer.from(b);
  return ba.length === bb.length && timingSafeEqual(ba, bb);
}

// ── BYOB credential encryption (matches PHP CredentialEncryptor) ──────────────
// AES-256-GCM. Key = HKDF-SHA256(ikm = secret, salt = 32 zero bytes,
// info = "fluxfiles-byob-enc", len = 32). Blob = base64(nonce[12] | ct | tag[16]).
const BYOB_INFO = 'fluxfiles-byob-enc';
const NONCE_LEN = 12;
const TAG_LEN = 16;

/**
 * Derive the BYOB key exactly like PHP `hash_hkdf('sha256', $secret, 32, $info)`.
 * PHP's empty HKDF salt means "HashLen zero bytes" (RFC 5869), so we pass an
 * explicit 32-byte zero salt — passing an empty Buffer to Node's hkdfSync would
 * NOT match.
 */
function deriveByobKey(secret: string): Buffer {
  const salt = Buffer.alloc(32, 0);
  const dk = hkdfSync('sha256', Buffer.from(secret, 'utf8'), salt, Buffer.from(BYOB_INFO, 'utf8'), 32);
  return Buffer.from(dk);
}

export function encryptByob(config: unknown, secret: string): string {
  const key = deriveByobKey(secret);
  const nonce = randomBytes(NONCE_LEN);
  const cipher = createCipheriv('aes-256-gcm', key, nonce, { authTagLength: TAG_LEN });
  // JSON.stringify does not escape "/", matching PHP's JSON_UNESCAPED_SLASHES.
  const plaintext = JSON.stringify(config);
  const ct = Buffer.concat([cipher.update(plaintext, 'utf8'), cipher.final()]);
  const tag = cipher.getAuthTag();
  return Buffer.concat([nonce, ct, tag]).toString('base64');
}

export function decryptByob<T = unknown>(blob: string, secret: string): T {
  const key = deriveByobKey(secret);
  const raw = Buffer.from(blob, 'base64');
  if (raw.length < NONCE_LEN + TAG_LEN + 1) {
    throw new Error('FluxFiles: invalid BYOB credential blob');
  }
  const nonce = raw.subarray(0, NONCE_LEN);
  const tag = raw.subarray(raw.length - TAG_LEN);
  const ct = raw.subarray(NONCE_LEN, raw.length - TAG_LEN);
  const decipher = createDecipheriv('aes-256-gcm', key, nonce, { authTagLength: TAG_LEN });
  decipher.setAuthTag(tag);
  const pt = Buffer.concat([decipher.update(ct), decipher.final()]).toString('utf8');
  return JSON.parse(pt) as T;
}

File: packages/python/src/fluxfiles_token/crypto.py
"""BYOB credential encryption — matches PHP `CredentialEncryptor` and
`packages/node/src/crypto.ts` byte-for-byte.

AES-256-GCM. Key = HKDF-SHA256(ikm = secret, salt = 32 zero bytes,
info = "fluxfiles-byob-enc", len = 32). Blob = base64(nonce[12] | ct | tag[16]).
"""

import base64
import json
import os
from typing import Any

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.hazmat.primitives.kdf.hkdf import HKDF

from .exceptions import FluxFilesByobError

_BYOB_INFO = b"fluxfiles-byob-enc"
_NONCE_LEN = 12
_TAG_LEN = 16
_KEY_LEN = 32


def derive_byob_key(secret: str) -> bytes:
    """Derive the BYOB key exactly like PHP `hash_hkdf('sha256', $secret, 32, $info)`.

    PHP's empty HKDF salt means "HashLen zero bytes" (RFC 5869), so we pass an
    explicit 32-byte zero salt. `cryptography`'s HKDF already defaults an
    omitted salt to the same RFC-5869-correct value, but we pin it explicitly
    anyway so this stays self-documenting and immune to any future change in
    that default — this single line is the crux of cross-language parity.
    """
    hkdf = HKDF(algorithm=hashes.SHA256(), length=_KEY_LEN, salt=b"\x00" * 32, info=_BYOB_INFO)
    return hkdf.derive(secret.encode("utf-8"))


def encrypt_byob(config: dict[str, Any], secret: str) -> str:
    """Encrypt a disk config into a base64 blob: nonce(12) || ciphertext || tag(16)."""
    key = derive_byob_key(secret)
    nonce = os.urandom(_NONCE_LEN)
    aesgcm = AESGCM(key)
    # json.dumps never escapes "/", matching PHP's JSON_UNESCAPED_SLASHES.
    plaintext = json.dumps(config, ensure_ascii=False).encode("utf-8")
    # AESGCM.encrypt() returns ciphertext || tag already concatenated (tag is
    # always the trailing 16 bytes) — no manual tag-splicing needed.
    ct_and_tag = aesgcm.encrypt(nonce, plaintext, None)
    return base64.b64encode(nonce + ct_and_tag).decode("ascii")


def decrypt_byob(blob: str, secret: str) -> dict[str, Any]:
    """Decrypt a base64 blob back into a disk config dict. Raises
    FluxFilesByobError on a malformed blob, wrong secret, or tampered data."""
    key = derive_byob_key(secret)
    try:
        raw = base64.b64decode(blob, validate=True)
    except Exception as exc:
        raise FluxFilesByobError("FluxFiles: invalid BYOB credential blob") from exc

    if len(raw) < _NONCE_LEN + _TAG_LEN + 1:
        raise FluxFilesByobError("FluxFiles: invalid BYOB credential blob")

    nonce = raw[:_NONCE_LEN]
    ciphertext_and_tag = raw[_NONCE_LEN:]
    aesgcm = AESGCM(key)
    try:
        # AESGCM.decrypt() expects ciphertext || tag concatenated — exactly the
        # wire format above, so no re-slicing beyond splitting off the nonce.
        plaintext = aesgcm.decrypt(nonce, ciphertext_and_tag, None)
    except Exception as exc:
        raise FluxFilesByobError(
            "FluxFiles: BYOB credential decryption failed — token may be tampered"
        ) from exc

    config = json.loads(plaintext.decode("utf-8"))
    if not isinstance(config, dict):
        raise FluxFilesByobError("FluxFiles: invalid BYOB credential format")
    return config
